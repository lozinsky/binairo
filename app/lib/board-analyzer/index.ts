export * from './analyze-board-by-kind';
export * from './analyze-board';
