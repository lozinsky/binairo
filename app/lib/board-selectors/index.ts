export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
