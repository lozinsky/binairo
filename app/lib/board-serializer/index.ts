export * from './board-serializer-collection';
