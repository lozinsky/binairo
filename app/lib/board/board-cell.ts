export type BoardCellKind = 'fixed' | 'regular';

export type BoardCellKindRaw = 0 | 1;

export type BoardCellState = 'B' | 'E' | 'R';

export type BoardCellStateRaw = 0 | 1 | 2;

export function isBoardCellKindRaw(value: unknown): value is BoardCellKindRaw {
  return typeof value === 'number' && value in BOARD_CELL_KIND_BY_RAW;
}

export function isBoardCellStateRaw(value: unknown): value is BoardCellStateRaw {
  return typeof value === 'number' && value in BOARD_CELL_STATE_BY_RAW;
}

const NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellState>> = {
  B: 'E',
  E: 'R',
  R: 'B',
};

const BALANCE_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, number>> = {
  B: -1,
  E: 0,
  R: 1,
};

const RAW_BY_BOARD_CELL_KIND: Readonly<Record<BoardCellKind, BoardCellKindRaw>> = {
  fixed: 0,
  regular: 1,
};

const BOARD_CELL_KIND_BY_RAW: Readonly<Record<BoardCellKindRaw, BoardCellKind>> = {
  0: 'fixed',
  1: 'regular',
};

const RAW_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellStateRaw>> = {
  B: 0,
  E: 2,
  R: 1,
};

const BOARD_CELL_STATE_BY_RAW: Readonly<Record<BoardCellStateRaw, BoardCellState>> = {
  0: 'B',
  1: 'R',
  2: 'E',
};

export type BoardCellValue = readonly [kind: BoardCellKindRaw, state: BoardCellStateRaw];

export class BoardCell {
  get balance() {
    return BALANCE_BY_BOARD_CELL_STATE[this.#state];
  }

  get isEmpty() {
    return this.#state === 'E';
  }

  get isFilled() {
    return !this.isEmpty;
  }

  get isFixed() {
    return this.#kind === 'fixed';
  }

  get state() {
    return this.#state;
  }

  readonly #kind: BoardCellKind;
  readonly #state: BoardCellState;

  constructor(kind: BoardCellKind, state: BoardCellState) {
    this.#kind = kind;
    this.#state = state;
  }

  static create(state: BoardCellState) {
    return new this('regular', state);
  }

  static from(value: BoardCellValue) {
    return new this(BOARD_CELL_KIND_BY_RAW[value[0]], BOARD_CELL_STATE_BY_RAW[value[1]]);
  }

  equals(other: BoardCell) {
    return this.#state === other.#state;
  }

  next() {
    return this.isFixed ? this : new BoardCell(this.#kind, NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE[this.#state]);
  }

  toFixed() {
    return new BoardCell('fixed', this.#state);
  }

  valueOf(): BoardCellValue {
    return [RAW_BY_BOARD_CELL_KIND[this.#kind], RAW_BY_BOARD_CELL_STATE[this.#state]];
  }
}

export function isBoardCellValue(value: unknown): value is BoardCellValue {
  if (Array.isArray(value) && value.length === 2) {
    return isBoardCellKindRaw(value[0]) && isBoardCellStateRaw(value[1]);
  }

  return false;
}
