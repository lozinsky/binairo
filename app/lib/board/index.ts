export * from './board-cell';
export * from './board-line';
export * from './board';
