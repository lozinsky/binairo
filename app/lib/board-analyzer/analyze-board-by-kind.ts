import type { Board, BoardOrientation } from '~/lib/board';

import {
  selectFilledImbalancedLine,
  selectLineThatCanBeEqualToAnother,
  selectLineThatIncludesAnother,
  selectLineThatIsEqualToOthers,
  selectMiddleCellBetweenTwoIdentical,
  selectNextCellAfterTwoIdentical,
  selectThreeOrMoreIdenticalSequentialCells,
  selectUnfilledBalancedLine,
} from '~/lib/board-selectors';
import { MatrixSelection, type MatrixSelectionPosition } from '~/lib/matrix';
import { type Random, shuffle } from '~/shared/random';

export interface BoardAnalyzerReview {
  readonly payload: BoardAnalyzerReviewPayload;
  readonly reason: BoardAnalyzerReviewReason;
}

export type BoardAnalyzerReviewKind = 'correction' | 'suggestion';

export interface BoardAnalyzerReviewPayload {
  readonly orientation: BoardOrientation;
  readonly positions: readonly MatrixSelectionPosition[];
}

export type BoardAnalyzerReviewReason =
  | 'line-can-be-equal-to-another'
  | 'line-includes-another'
  | 'line-is-equal-to-others'
  | 'line-is-filled-and-imbalanced'
  | 'line-is-unfilled-and-balanced'
  | 'middle-cell-between-two-identical'
  | 'next-cell-after-two-identical'
  | 'three-or-more-identical-sequential-cells';

const BOARD_ANALYZER_REVIEW_REASONS_BY_BOARD_ANALYZER_REVIEW_KIND: Readonly<
  Record<BoardAnalyzerReviewKind, readonly BoardAnalyzerReviewReason[]>
> = {
  correction: ['line-is-filled-and-imbalanced', 'line-is-equal-to-others', 'three-or-more-identical-sequential-cells'],
  suggestion: [
    'line-is-unfilled-and-balanced',
    'line-can-be-equal-to-another',
    'line-includes-another',
    'middle-cell-between-two-identical',
    'next-cell-after-two-identical',
  ],
};

const SELECT_BY_BOARD_ANALYZER_REVIEW_REASON: Readonly<
  Record<BoardAnalyzerReviewReason, (target: Board) => BoardAnalyzerReviewPayload | undefined>
> = {
  'line-can-be-equal-to-another'(target) {
    const payload = selectLineThatCanBeEqualToAnother(target);

    if (payload === undefined) {
      return;
    }

    const cells = [...Array.from(payload.line), ...Array.from(payload.another)];

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'line-includes-another'(target) {
    const payload = selectLineThatIncludesAnother(target);

    if (payload === undefined) {
      return;
    }

    const cells = [...Array.from(payload.line), ...Array.from(payload.another)];

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'line-is-equal-to-others'(target) {
    const payload = selectLineThatIsEqualToOthers(target);

    if (payload === undefined) {
      return;
    }

    const cells = [...Array.from(payload.line), ...payload.others.flatMap((line) => Array.from(line))];

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'line-is-filled-and-imbalanced'(target) {
    const payload = selectFilledImbalancedLine(target);

    if (payload === undefined) {
      return;
    }

    const cells = Array.from(payload.line);

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'line-is-unfilled-and-balanced'(target) {
    const payload = selectUnfilledBalancedLine(target);

    if (payload === undefined) {
      return;
    }

    const cells = Array.from(payload.line);

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'middle-cell-between-two-identical'(target) {
    const payload = selectMiddleCellBetweenTwoIdentical(target);

    if (payload === undefined) {
      return;
    }

    const cells = [payload.cell];

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'next-cell-after-two-identical'(target) {
    const payload = selectNextCellAfterTwoIdentical(target);

    if (payload === undefined) {
      return;
    }

    const cells = [payload.cell];

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },

  'three-or-more-identical-sequential-cells'(target) {
    const payload = selectThreeOrMoreIdenticalSequentialCells(target);

    if (payload === undefined) {
      return;
    }

    const cells = payload.cells;

    return {
      orientation: payload.orientation,
      positions: MatrixSelection.collect(target, cells).valueOf(),
    };
  },
};

export function analyzeBoardByKind(
  kind: BoardAnalyzerReviewKind,
  target: Board,
  random: Random,
): BoardAnalyzerReview | undefined {
  const reasons = shuffle(BOARD_ANALYZER_REVIEW_REASONS_BY_BOARD_ANALYZER_REVIEW_KIND[kind], random);

  for (const reason of reasons) {
    const select = SELECT_BY_BOARD_ANALYZER_REVIEW_REASON[reason];
    const payload = select(target);

    if (payload === undefined) {
      continue;
    }

    return { payload, reason };
  }

  return;
}
