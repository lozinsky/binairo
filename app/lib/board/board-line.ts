import { DoubleMatrixRotation, type MatrixLine, type MatrixRotatable } from '~/lib/matrix';
import { expectToBeDefined } from '~/shared/expect';

import { BoardCell, BoardCellState, type BoardCellValue } from './board-cell';

export type BoardLineValue = readonly BoardCellValue[];

export class BoardLine implements Iterable<BoardCell>, MatrixLine<BoardCell>, MatrixRotatable<BoardLine> {
  readonly #cells: readonly BoardCell[];

  constructor(cells: readonly BoardCell[]) {
    this.#cells = cells;
  }

  static blank(size: number) {
    return new this(Array.from({ length: size }, () => BoardCell.create(BoardCellState.E)));
  }

  static create(states: readonly BoardCellState[]) {
    return new this(states.map((state) => BoardCell.create(state)));
  }

  static from(value: BoardLineValue) {
    return new this(value.map((value) => BoardCell.from(value)));
  }

  [Symbol.iterator]() {
    return this.#cells[Symbol.iterator]();
  }

  at(index: number) {
    return this.#cells.at(index);
  }

  entries() {
    return this.#cells.entries();
  }

  equals(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((cell, index) => cell.equals(expectToBeDefined(other.at(index))));
  }

  includes(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((a, index) => {
      const b = expectToBeDefined(other.at(index));

      return b.isEmpty ? true : a.equals(b);
    });
  }

  indexOf(cell: BoardCell) {
    return this.#cells.indexOf(cell);
  }

  keys() {
    return this.#cells.keys();
  }

  lengthOf(state: BoardCellState) {
    return this.#cells.reduce((length, cell) => (cell.state === state ? length + 1 : length), 0);
  }

  rotate() {
    return new BoardLine(this.#cells.toReversed());
  }

  slice(start: number, end?: number) {
    return this.#cells.slice(start, end);
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<BoardLine>(this);
  }

  valueOf(): BoardLineValue {
    return this.#cells.map((cell) => cell.valueOf());
  }

  get balance() {
    return this.#cells.reduce((balance, cell) => balance + cell.balance, 0);
  }

  get isEmpty() {
    return this.#cells.every((cell) => cell.isEmpty);
  }

  get isFilled() {
    return this.#cells.every((cell) => cell.isFilled);
  }

  get length() {
    return this.#cells.length;
  }

  get progress() {
    return this.#cells.reduce((progress, cell) => (cell.isFilled ? progress + 1 : progress), 0) / this.length;
  }
}
