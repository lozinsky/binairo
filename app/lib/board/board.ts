import {
  DoubleMatrixRotation,
  type Matrix,
  type MatrixReversible,
  type MatrixRotatable,
  type MatrixSelection,
  QuadrupleMatrixRotation,
} from '~/lib/matrix';
import { assert } from '~/shared/assert';
import { expectToBeDefined } from '~/shared/expect';

import type { BoardCell, BoardCellState } from './board-cell';

import { BoardLine, type BoardLineValue, isBoardLineValue } from './board-line';

export type BoardOrientation = 'landscape' | 'portrait';

const NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION: Readonly<Record<BoardOrientation, BoardOrientation>> = {
  landscape: 'portrait',
  portrait: 'landscape',
};

export type BoardValue = readonly BoardLineValue[];

export class Board implements Iterable<BoardLine>, Matrix<BoardLine>, MatrixReversible<Board>, MatrixRotatable<Board> {
  get length() {
    return this.#lines.length;
  }

  get orientation() {
    return this.#orientation;
  }

  get progress() {
    return this.#lines.reduce((progress, line) => progress + line.progress, 0) / this.#lines.length;
  }

  readonly #lines: readonly BoardLine[];
  readonly #orientation: BoardOrientation;

  constructor(orientation: BoardOrientation, lines: readonly BoardLine[]) {
    for (const line of lines) {
      assert(line.length === lines.length, 'Expected line length to be equal to board size');
    }

    this.#lines = lines;
    this.#orientation = orientation;
  }

  static blank(size: number) {
    return new this(
      'portrait',
      Array.from({ length: size }, () => BoardLine.blank(size)),
    );
  }

  static create(states: ReadonlyArray<readonly BoardCellState[]>) {
    return new this(
      'portrait',
      states.map((states) => BoardLine.create(states)),
    );
  }

  static from(value: BoardValue) {
    return new this(
      'portrait',
      value.map((value) => BoardLine.from(value)),
    );
  }

  at(index: number) {
    return this.#lines.at(index);
  }

  entries() {
    return this.#lines.entries();
  }

  replace(index: number, callback: (line: BoardLine) => BoardLine) {
    return new Board(this.#orientation, this.#lines.with(index, callback(expectToBeDefined(this.#lines.at(index)))));
  }

  replaceBy(selection: MatrixSelection, callback: (cell: BoardCell) => BoardCell) {
    const content = selection.execute(this);
    const lines = this.#lines.map((line) => {
      const cells = Array.from(line);

      for (const cell of content) {
        const index = line.indexOf(cell);

        if (index === -1) {
          continue;
        }

        cells[index] = callback(expectToBeDefined(cells[index]));
      }

      return new BoardLine(cells);
    });

    return new Board(this.#orientation, lines);
  }

  reverse() {
    return new Board(this.#orientation, this.#lines.toReversed());
  }

  rotate() {
    const orientation = NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION[this.#orientation];

    if (this.#lines.length === 0) {
      return new Board(orientation, []);
    }

    const lines: BoardLine[] = [];

    for (const index of expectToBeDefined(this.#lines[0]).keys()) {
      const cells: BoardCell[] = [];

      for (const line of this.#lines) {
        cells.push(expectToBeDefined(line.at(index)));
      }

      lines.push(new BoardLine(cells));
    }

    return new Board(orientation, lines);
  }

  slice(start: number, end?: number) {
    return this.#lines.slice(start, end);
  }

  [Symbol.iterator]() {
    return this.#lines[Symbol.iterator]();
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<Board>(this);
  }

  toQuadrupleRotation() {
    return QuadrupleMatrixRotation.run<Board>(this);
  }

  valueOf(): BoardValue {
    return this.#lines.map((line) => line.valueOf());
  }
}

export function isBoardValue(value: unknown): value is BoardValue {
  return Array.isArray(value) && value.every(isBoardLineValue);
}
