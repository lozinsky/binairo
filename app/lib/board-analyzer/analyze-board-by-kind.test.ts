import { expect, test } from 'vitest';

import { Board } from '~/lib/board';
import { Random } from '~/shared/random';

import type { BoardAnalyzerReviewKind } from './analyze-board-by-kind';

import { analyzeBoardByKind } from './analyze-board-by-kind';

test.each<[BoardAnalyzerReviewKind, Board]>([
  [
    'correction',
    Board.create([
      ['E', 'B', 'E', 'B'],
      ['E', 'E', 'B', 'E'],
      ['B', 'E', 'R', 'E'],
      ['B', 'R', 'B', 'R'],
    ]),
  ],
  [
    'correction',
    Board.create([
      ['E', 'B', 'E', 'B'],
      ['B', 'E', 'B', 'E'],
      ['B', 'E', 'R', 'E'],
      ['B', 'R', 'B', 'R'],
    ]),
  ],
  [
    'suggestion',
    Board.create([
      ['E', 'B', 'E', 'B'],
      ['E', 'E', 'B', 'E'],
      ['B', 'E', 'R', 'E'],
      ['B', 'R', 'B', 'R'],
    ]),
  ],
  [
    'correction',
    Board.create([
      ['B', 'B', 'R', 'R', 'E', 'R'],
      ['R', 'R', 'B', 'R', 'E', 'E'],
      ['E', 'E', 'E', 'B', 'R', 'E'],
      ['E', 'E', 'R', 'B', 'R', 'E'],
      ['R', 'E', 'B', 'R', 'E', 'E'],
      ['R', 'R', 'E', 'E', 'R', 'B'],
    ]),
  ],
  [
    'correction',
    Board.create([
      ['B', 'B', 'R', 'R', 'E', 'R'],
      ['R', 'R', 'B', 'R', 'E', 'E'],
      ['E', 'E', 'E', 'B', 'R', 'E'],
      ['R', 'E', 'R', 'B', 'R', 'E'],
      ['R', 'E', 'B', 'R', 'E', 'E'],
      ['R', 'R', 'E', 'E', 'R', 'B'],
    ]),
  ],
  [
    'suggestion',
    Board.create([
      ['B', 'B', 'R', 'R', 'E', 'R'],
      ['R', 'R', 'B', 'R', 'E', 'E'],
      ['E', 'E', 'E', 'B', 'R', 'E'],
      ['E', 'E', 'R', 'B', 'R', 'E'],
      ['R', 'E', 'B', 'R', 'E', 'E'],
      ['R', 'R', 'E', 'E', 'R', 'B'],
    ]),
  ],
])('analyzes board by kind', (kind, target) => {
  expect(analyzeBoardByKind(kind, target, Random.stable())).toMatchSnapshot();
});
