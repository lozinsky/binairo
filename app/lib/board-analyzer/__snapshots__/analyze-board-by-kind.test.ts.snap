// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`analyzes board by kind 1`] = `undefined`;

exports[`analyzes board by kind 2`] = `
{
  "payload": {
    "orientation": "landscape",
    "positions": [
      {
        "x": 0,
        "y": 1,
      },
      {
        "x": 0,
        "y": 2,
      },
      {
        "x": 0,
        "y": 3,
      },
    ],
  },
  "reason": "three-or-more-identical-sequential-cells",
}
`;

exports[`analyzes board by kind 3`] = `
{
  "payload": {
    "orientation": "landscape",
    "positions": [
      {
        "x": 0,
        "y": 1,
      },
    ],
  },
  "reason": "next-cell-after-two-identical",
}
`;

exports[`analyzes board by kind 4`] = `undefined`;

exports[`analyzes board by kind 5`] = `
{
  "payload": {
    "orientation": "landscape",
    "positions": [
      {
        "x": 0,
        "y": 3,
      },
      {
        "x": 0,
        "y": 4,
      },
      {
        "x": 0,
        "y": 5,
      },
    ],
  },
  "reason": "three-or-more-identical-sequential-cells",
}
`;

exports[`analyzes board by kind 6`] = `
{
  "payload": {
    "orientation": "portrait",
    "positions": [
      {
        "x": 4,
        "y": 0,
      },
    ],
  },
  "reason": "next-cell-after-two-identical",
}
`;
